### PR TITLE
MULE-18020: Fix reconnection for Paged Operations that fail after the first page.

### PR DESCRIPTION
**Pull request by SebaElizalde**

### CODE REVIEW

File: modules/extensions-spring-support/src/test/java/org/mule/test/module/extension/reconnection/ReconnectionTestCase.java [hunk 1]
```diff
@@ -106,13 +108,13 @@ public void getInlineRetryPolicyTemplate() throws Exception {
...
-    ReconnectableConnection firstPage = iterator.next();
-    assertThat("Connection was not disconnected.", firstPage.getDisconnectCalls(), is(1));
+    iterator.next();
+    assertThat("Connection was not disconnected.", disconnectCalls, is(1));
     assertThat("Paging provider was not closed.", closePagingProviderCalls, is(1));
   }
 
-  @Test
-  public void doNotReconnectAfterOtherExceptionOnFirstPage() {
+  @Test(expected = IllegalArgumentException.class)
```
Comment (marianogonzalez) [inline modules/extensions-spring-support/src/test/java/org/mule/test/module/extension/reconnection/ReconnectionTestCase.java]: use the ExpectedException rule and assert on the cause being THE SAME instance you are expecting and the actual exception being of the correct type

File: modules/extensions-spring-support/src/test/java/org/mule/test/module/extension/reconnection/ReconnectionTestCase.java [hunk 2]
```diff
@@ -123,18 +125,83 @@ public void doNotReconnectAfterOtherExceptionOnFirstPage() {
       assertThat(e.getMessage(), is("An illegal argument was received."));
       assertThat("Paging provider was not closed.", closePagingProviderCalls, is(1));
       assertThat("Connection was disconnected.", disconnectCalls, is(0));
+      throw e.getCause();
```
Comment (marianogonzalez) [inline modules/extensions-spring-support/src/test/java/org/mule/test/module/extension/reconnection/ReconnectionTestCase.java]: replace all of this with the expectedException rule

File: modules/extensions-spring-support/src/test/java/org/mule/test/module/extension/reconnection/ReconnectionTestCase.java [hunk 3]
```diff
@@ -123,18 +125,83 @@ public void doNotReconnectAfterOtherExceptionOnFirstPage() {
...
+  @Test(expected = IllegalArgumentException.class)
+  public void doNotReconnectAfterOtherExceptionOnSecondPage() throws Exception {
+    resetCounters();
+    Iterator<ReconnectableConnection> iterator;
+    try {
+      iterator = getCursor("pagedOperation", 2, VALIDATION);
+      iterator.next();
+      iterator.next();
+    } catch (Exception e) {
+      assertThat(e, instanceOf(IllegalArgumentException.class));
```
Comment (marianogonzalez) [inline modules/extensions-spring-support/src/test/java/org/mule/test/module/extension/reconnection/ReconnectionTestCase.java]: same as before

File: modules/extensions-spring-support/src/test/java/org/mule/test/module/extension/reconnection/ReconnectionTestCase.java [hunk 4]
```diff
@@ -123,18 +125,83 @@ public void doNotReconnectAfterOtherExceptionOnFirstPage() {
...
 
+  @Test(expected = ModuleException.class)
+  public void stickyConnectionIsNotReconnectedDuringConnectionExceptionOnSecondPage() throws Exception {
+    resetCounters();
+    try {
+      Iterator<ReconnectableConnection> iterator = getCursor("stickyPagedOperation", 2, CONNECTIVITY);
+      iterator.next();
+      iterator.next();
+    } catch (Exception e) {
+      assertThat(e, instanceOf(ModuleException.class));
```
Comment (marianogonzalez) [inline modules/extensions-spring-support/src/test/java/org/mule/test/module/extension/reconnection/ReconnectionTestCase.java]: same as before

File: modules/extensions-spring-support/src/test/java/org/mule/test/module/extension/reconnection/ReconnectionTestCase.java [hunk 5]
```diff
@@ -123,18 +125,83 @@ public void doNotReconnectAfterOtherExceptionOnFirstPage() {
...
+
+  @Test(expected = IllegalArgumentException.class)
+  public void stickyConnectionIsNotReconnectedDuringOtherExceptionOnSecondPage() throws Exception {
+    resetCounters();
+    try {
+      Iterator<ReconnectableConnection> iterator = getCursor("stickyPagedOperation", 2, VALIDATION);
+      iterator.next();
+      iterator.next();
+    } catch (Exception e) {
+      assertThat(e, instanceOf(IllegalArgumentException.class));
```
Comment (marianogonzalez) [inline modules/extensions-spring-support/src/test/java/org/mule/test/module/extension/reconnection/ReconnectionTestCase.java]: same as before

File: modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/runtime/streaming/PagingProviderProducer.java [hunk 6]
```diff
@@ -91,27 +101,58 @@ public int getSize() {
...
     } finally {
       safely(connectionSupplier::close, e -> LOGGER.debug("Found exception closing the connection supplier", e));
     }
   }
 
+  private void handleConnectionException(ConnectionException connectionException, ConnectionSupplier connectionSupplier) {
+    Optional<String> exceptionConfigName =
+        executionContext.getConfiguration().map(config -> ((ConfigurationInstance) config).getName());
+    if (isTransactional()) {
+      connectionException.addInfo("wasTransactional", true);
```
Comment (marianogonzalez) [inline modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/runtime/streaming/PagingProviderProducer.java]: Define constant in org.mule.runtime.module.extension.internal.ExtensionProperties. Also the key should start with the proper prefix.

File: modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/runtime/streaming/PagingProviderProducer.java [hunk 7]
```diff
@@ -91,27 +101,58 @@ public int getSize() {
...
+    }
+    if (isFirstPage) {
+      exceptionConfigName.ifPresent(name -> connectionException.addInfo("operationConfigName", name));
```
Comment (marianogonzalez) [inline modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/runtime/streaming/PagingProviderProducer.java]: same

File: modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/util/ReconnectionUtils.java [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+    return validateConnectionException(connectionException.get(), contextConfigName.orElse(null));
+  }
+
+  private static boolean validateConnectionException(ConnectionException connectionException, String contextConfigName) {
+    Boolean wasTransactional = (Boolean) connectionException.getInfo().get("wasTransactional");
+    if (wasTransactional != null && wasTransactional) {
+      return false;
+    }
+    Object operationConfigName = connectionException.getInfo().get("operationConfigName");
+    if (operationConfigName != null && contextConfigName != null) {
```
Comment (marianogonzalez) [inline modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/util/ReconnectionUtils.java]: return Objects.equals(operaetionConfigName, contextConfigName)

File: modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/util/ReconnectionUtils.java [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+    if (isTransactionActive()) {
+      Transaction tx = TransactionCoordination.getInstance().getTransaction();
+
+      return !tx.hasResource(new ExtensionTransactionKey(context.getConfiguration().get()));
+    }
+
+    return validateConnectionException(connectionException.get(), contextConfigName.orElse(null));
+  }
+
+  private static boolean validateConnectionException(ConnectionException connectionException, String contextConfigName) {
```
Comment (marianogonzalez) [inline modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/util/ReconnectionUtils.java]: a boolean method should not start with validate. Give this a more meaningful name and add a comment explaining why this is necessary

File: modules/extensions-spring-support/src/test/java/org/mule/test/module/extension/reconnection/ReconnectionTestCase.java (hunk 1 above)
Comment (SebaElizalde) [reply modules/extensions-spring-support/src/test/java/org/mule/test/module/extension/reconnection/ReconnectionTestCase.java]: I tried the approach with ExpectedException, but it does not allow me to assert other things after the exception is thrown (like the paging provider closing or the disconnection calls).

**Comment by marianogonzalez** [comment]
--validate

**Pull request merged**